lib/update/cluster/phases/init.go

```diff
@@ -194,34 +194,64 @@ func (p *updatePhaseInit) Rollback(ctx context.Context) error {
 			return trace.Wrap(err)
 		}
 	}
+	err := p.restoreRPCCredentials()
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
 	if err := p.removeConfiguredPackages(); err != nil {
 		return trace.Wrap(err)
 	}
 	return nil
 }
 
-func (p *updatePhaseInit) initRPCCredentials() error {
-	// FIXME: the secrets package is currently only generated once.
-	// Even though the package is generated with some time buffer in advance,
-	// we need to make sure if the existing package needs to be rotated (i.e.
-	// as expiring soon).
-	// This will ether need to generate a new package version and then the
-	// problem becomes how the agents will know the name of the package.
-	// Or, the package version is recycled and then we need to make sure
-	// to restart the cluster controller (gravity-site) to make sure it has
-	// reloaded its copy of the credentials.
-	// See: https://github.com/gravitational/gravity/issues/3607.
-	pkg, err := rpc.InitCredentials(p.Packages)
-	if err != nil && !trace.IsAlreadyExists(err) {
+// updateRPCCredentials rotates the RPC credentials used for install/expand/leave operations.
+func (p *updatePhaseInit) updateRPCCredentials() error {
+	// This assumes that the cluster controller Pods are eventually restarted
+	// by the upcoming phase for these changes to take effect.
+	//
+	// Currently the upgrade short-circuits the application-only upgrades by not
+	// including the init phase so this is safe.
+	//
+	// Keep it in mind for future changes.
+	// See https://github.com/gravitational/gravity/issues/3607 for more details when we had
+	// to be careful about it previously.
+	p.Info("Update RPC credentials")
+	if err := p.backupRPCCredentials(); err != nil {
 		return trace.Wrap(err)
 	}
+	loc, err := rpc.UpsertCredentials(p.Packages)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	p.WithField("package", loc.String()).Info("Update RPC credentials.")
+	return nil
+}
 
-	if trace.IsAlreadyExists(err) {
-		p.Info("RPC credentials already initialized.")
-	} else {
-		p.Infof("Initialized RPC credentials: %v.", pkg)
+func (p *updatePhaseInit) backupRPCCredentials() error {
+	p.Info("Backup RPC credentials")
+	env, rc, err := rpc.LoadCredentialsData(p.Packages)
+	if err != nil {
+		return trace.Wrap(err)
 	}
+	defer rc.Close()
+	_, err = p.Packages.CreatePackage(rpcBackupPackage, rc, pack.WithLabels(env.RuntimeLabels))
```

Comment:
Shouldn't this be "upsert"? Otherwise on the second upgrade this will fail because the package will already be there?

Reply:
I've already made a change to assign it the corresponding operation label. Ideally, packages marked for the specific operation need to be purged post-operation. I'll make it into upsert though.

Reply:
Yeah I think upsert is better anyway just so the phase is reentrable e.g. if reexecuted multiple times.